### PR TITLE
Add first batch of submissions tests, refactor backend code to have more error handling / validation

### LINKED ISSUE #2109: Version numbering should reset on submission deletion
As a result on #2024, version numbering of submissions no longer resets after the deletion of a submission.

e.g. previously, creating submissions 1 and 2, deleting both, and then creating a new submission would result in version 1 again.

Now, it results in version 3.

To remedy, add the necessary logic to AUD's `update_latest_submissions`, which is called on `after_destroy` of a submission.

### PR DESCRIPTION
**Pull request by 20wildmanj**
## Description

1. Remove the `#show` route, which is unused, from `routes.rb` and `submissions_controller.rb`
2. Refactor `#new` to have better and correct error-checking so that a 500 error wouldn't occur when a CUD ID didn't exist in the list of CUD IDs provided to the action
3. Add validation of params / error-checking in `#create` on invalid submission creation
4. Add similar error-checking to `#update`
6. Update the submission destroy interface to be more user-friendly and make more sense, by disabling the button to destroy the submission unless the instructor has checked the box indicating the read the disclaimer and still wish to destroy the submission. Made buttons and checkbox more descriptive as well.

Before: Instructor could still click button to destroy the submission without checking the box but would get an "error" back.

Screenshot 2024-03-08 at 6 22 03 PM
Screenshot 2024-03-08 at 6 22 08 PM

After: Instructor can't click button without having checked the box first.

Screenshot 2024-03-08 at 6 20 49 PM
Screenshot 2024-03-08 at 6 21 00 PM

7. Add tests for:
- new
- create
- edit

(also try to fix flaky assessment_spec test by giving capybara expect more wait time)

8. Fix version number not decreasing when assessment is deleted
- add method `delete_version_number` to decrease version number when submission is destroyed, and update `update_latest_submission` so that it finds the latest submission based on creation date (and checks with version number)

## Motivation and Context

Fixes #2109 

## How Has This Been Tested?

- Try to create submission for CUD with ids via url e.g. `http://localhost:3000/courses/AutoPopulated/assessments/lab5/submissions/new?course_user_datum_id=29,30`, see that it renders correctly
- try to do some with non-existent CUD id, see that error appears

Screenshot 2024-03-08 at 7 12 49 PM

- verify that creating missing submissions still work
- Run through destroy submission flow, see that it works and makes sense
- Create submissions for a user (can be yourself), then destroy them, then create more submissions, see that submission number updates correctly
- Run tests locally and see that they pass

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting

### CODE REVIEW

File: app/models/assessment_user_datum.rb [hunk 1]
```diff
@@ -72,14 +72,17 @@ def update_latest_submission
...
 
-  # Calculate latest unignored submission (i.e. with max version and unignored)
+  # Calculate latest unignored submission (i.e. with latest creation date and unignored)
   def latest_submission!
-    if (max_version = Submission.where(assessment_id:,
-                                       course_user_datum_id:,
-                                       ignored: false).maximum(:version))
-      Submission.find_by(version: max_version, assessment_id:,
-                         course_user_datum_id:)
+    latest_creation_time = Submission.where(assessment_id:,
```
Comment (damianhxy) [inline app/models/assessment_user_datum.rb:77]: Is this a refactor or a bug fix?

File: app/models/assessment_user_datum.rb [hunk 2]
```diff
@@ -72,14 +72,17 @@ def update_latest_submission
...
+                                            course_user_datum_id:,
+                                            ignored: false).maximum(:created_at)
+    max_ver_submission = Submission.find_by(created_at: latest_creation_time, assessment_id:,
+                                            course_user_datum_id:)
+    if max_ver_submission && max_ver_submission.version != version_number
```
Comment (damianhxy) [inline app/models/assessment_user_datum.rb:82]: Likewise here

**Review by damianhxy — COMMENTED** [review]
Functionality looks good other than slightly unintuitive version numbering behavior

Would also like further elaboration on some logic changes

File: app/models/assessment_user_datum.rb (hunk 1 above)
Comment (20wildmanj) [reply app/models/assessment_user_datum.rb:77]: I'll revert this change

File: app/models/assessment_user_datum.rb [hunk 3]
```diff
@@ -246,6 +249,18 @@ def update_version_number
     self.version_number
   end
 
+  def delete_version_number
```
[the remark this answers is not in the document]
Comment (20wildmanj) [reply app/models/assessment_user_datum.rb:252]: done

**Review by damianhxy — APPROVED** [review]
LGTM

**Pull request merged**